Fixes Overpayment Calculation

Tapanito opened the pull request:
As title says, this fixes overpayment component calculation.

Before this fix, fee was taken of overpayments twice.


Furthermore, this PR adds unit tests to test the math logic for overpayments. The current approach of integration tests is not suitable to find such bugs, as it requires too much setup, and too specific conditions to surface such bugs.

Tapanito committed:
fixes overpayment interest calculation

Tapanito committed:
fixes formatting issues

Tapanito committed:
adds additional tests

Tapanito committed:
adds additional unit tests to cover math calculations

Tapanito committed:
removes unused methods

Tapanito committed:
fixes formatting

ximinez commented on src/test/app/LendingHelpers_test.cpp:
```diff
@@ -0,0 +1,551 @@
+#include <xrpl/protocol/SField.h>
+
+#include "test/jtx/Account.h"
+#include "test/jtx/amount.h"
```
Our convention is for all includes to be in angle brackets `<>` unless there's a good reason not to.

Tapanito replied:
Thanks Ed, I really need to clean up my clangd config to stop doing that...

ximinez commented on src/test/app/LendingHelpers_test.cpp:
```diff
@@ -0,0 +1,551 @@
+        PrettyAsset const IOU = issuer["IOU"];
+        int32_t const loanScale = 1;
+        auto const overpayment = Number{1'000};
+        auto const overpaymentInterestRate = TenthBips32{10'000};  // 10%
+        auto const overpaymentFeeRate = TenthBips32{10'000};       // 10%
+        auto const managementFeeRate = TenthBips16{10'000};        // 10%
```
This test might benefit from using different rate values. If they're all the same, you can't be sure that, e.g., the management fee was computed using the overpayment interest rate. 10%, 15%, and 20% would probably be fine.

Tapanito replied:
Good idea, changed overpayment fee to 50%

ximinez commented on src/test/app/LendingHelpers_test.cpp:
```diff
@@ -0,0 +1,551 @@
+
+        auto const testCases = std::vector<TestCase>{
+            {
+                .name = "On-time payment",
+                .principalOutstanding = Number{1'000},
+                .lateInterestRate = TenthBips32{10'000},  // 10%
+                .parentCloseTime =
+                    NetClock::time_point{NetClock::duration{3'000}},
+                .nextPaymentDueDate = 3'000,
```
Could you update this test case, or add another, so that `parentCloseTime` is _before_ `nextPaymentDueDate`, to verify we don't get negative interest?

Tapanito replied:
Done!

ximinez commented on src/test/app/LendingHelpers_test.cpp:
```diff
@@ -0,0 +1,551 @@
+                    NetClock::time_point{NetClock::duration{3'000}},
+                .startDate = 2'000,
+                .prevPaymentDate = 2'500,
+                .paymentInterval = 0,
```
This is a good test. Might be worth noting that in practice, in the loan transactions and objects, `paymentInterval` can never be 0.

Tapanito replied:
Yes, of course. I's one of those tests to ensure the system doesn't explode if somehow a zero values ends up sneaking in.

ximinez commented on src/test/app/LendingHelpers_test.cpp:
```diff
@@ -0,0 +1,551 @@
+    run() override
+    {
+        testLoanAccruedInterest();
+        testLoanLatePayementInterest();
```
Typo: `testLoanLatePaymentInterest()`.

Tapanito replied:
Fixed - Sausage Fingers :)

ximinez commented on src/xrpld/app/misc/detail/LendingHelpers.cpp:
```diff
@@ -1231,17 +1210,12 @@ computeOverpaymentComponents(
         [&]() {
-            Number const interest =
-                roundToAsset(asset, rawOverpaymentInterest, loanScale);
+            auto const interest = roundToAsset(
+                asset,
+                tenthBipsOfValue(overpayment, overpaymentInterestRate),
+                loanScale);
```
Nice optimization!

Tapanito replied:
Thanks! This is also a bug fix, prior to this, we calculated interest twice 😁

ximinez replied:
> Thanks! This is also a bug fix, prior to this, we calculated interest twice 😁

OMG. You're right!

ximinez reviewed (changes requested):
These changes are really minor. Overall this is a nice set of tests!

ximinez committed:
Merge branch 'ximinez/develop-nolending' into ximinez/lending-XLS-66-2

Tapanito committed:
Fix overpayment asserts (#6084)

ximinez committed:
Review feedback from @dangell7: early return & coverage

- Exclude LogicError lines in ApplyView.cpp (specifically directory
  operations) from code coverage.
- Replace the ability to set the next page on a new directory page with
  an assert, because nothing uses it right now.
- Early return with success for batch inner transactions in preflight2.

ximinez committed:
Merge branch 'ximinez/develop-nolending' into ximinez/lending-XLS-66-2

Tapanito committed:
Merge branch 'ximinez/lending-XLS-66-2' into tapanito/lending-fix-overpayment

Tapanito committed:
addresses review comments

Tapanito commented:
Thank you for reviewing the code Ed, I addressed your comments in this commit: https://github.com/XRPLF/rippled/pull/6087/commits/d66b299ebd8d33c7aa0f43001cfbba4cda07bebd

Tapanito committed:
formatting issues

Tapanito committed:
tests fullPaymentInterest

Tapanito committed:
removes unused code

Tapanito commented:
@ximinez , I added additional test for full payment interest, and removed some additional unused code

ximinez committed:
Test updates - show balances in runLoan()

ximinez committed:
Merge remote-tracking branch 'XRPLF/ximinez/develop-nolending' into ximinez/lending-XLS-66-2

* XRPLF/ximinez/develop-nolending:
  Revert "Implement Lending Protocol (unsupported) (5270)"
  Implement Lending Protocol (unsupported) (5270)

Tapanito committed:
Merge branch 'ximinez/lending-XLS-66-2' into tapanito/lending-fix-overpayment

ximinez committed:
MPTTester::operator() parameter should be std::int64_t

- Originally defined as uint64_t, but the testIssuerLoan() test called
  it with a negative number, causing an overflow to a very large number
  that in some circumstances could be silently cast back to an int64_t,
  but might not be. I believe this is UB, and we don't want to rely on
  that.

Tapanito committed:
Merge branch 'ximinez/lending-XLS-66-2' into tapanito/lending-fix-overpayment

Tapanito committed:
Merge branch 'ximinez/lending-XLS-66-ongoing' into tapanito/lending-fix-overpayment

Tapanito committed:
fixes formatting

ximinez reviewed (approved):
Nice work. Thanks for adding these finer grained tests.

Tapanito committed:
Merge branch 'ximinez/lending-XLS-66-ongoing' into tapanito/lending-fix-overpayment
